lib/heat.js

```
@@ -0,0 +1,293 @@
+    {
+      if(extra_headers.hasOwnProperty(key))
+      {
+        return_object.headers[key] = extra_headers[key];
+      }
+    }
+  }
+
+  return return_object;
+};
+
+
+
+//gets a list of all stacks for the given project/tenant
+//calls back with cb(error, stack_array)
+// the options object can be used to specify filters, e.g.:
+//{
+//    id: string,
+//    status: string,
+//    name: string,
+//    action: string,
+//    tenant: string,
+//    username: string,
+//    owner_id: string
+//}
+Heat.prototype.listStacks = function(options, cb)
+{
+  var self = this;
+  var args = OSUtils.getArgsWithCallback(this.listStacks.length, arguments);
+  options = args[0] || {};
+  cb = args[1];
+
+  var request_options = this.getRequestOptions('/stacks', true);
+  request_options.qs = options;
+  request_options.metricPath = 'remote-calls.heat.stacks.list';
+  request_options.validateStatus = true;
+  request_options.requireBodyObject = 'stacks';
+
+  this.request.get(request_options, function(error, response, body){
+    var stacks_array = [];
+    var n = 0;
+
+    if(error)
+    {
+      cb(error);
+      return;
+    }
+    //else
+
+    for (n = 0; n < body.stacks.length; n++)
+    {
+      stacks_array[n] = self.mangleObject('Stack', body.stacks[n]);
+    }
+
+    cb(null, stacks_array);
+  });
+}
+
+
+
+//creates a stack with the given name
+//calls back with cb(error, stack_object)
+// options:
+// {
+//   disable_rollback: boolean,
+//   environment: object,
+//   files: object,
+//   parameters: object,
+//   tags: string,
+//   template: object,
+//   template_url: string,
+//   timeout_mins: number
+//}
+//note: either template or template_url must be defined
+Heat.prototype.createStack = function(name, options, cb)
+{
+  var self = this;
+  var args = OSUtils.getArgsWithCallback(this.createStack.length, arguments);
+  name = args[0];
+  if (typeof name != 'string') {
+      throw new Error('heat.createStack: requires `name` (string) parameter');
```

Couple of questions on the param checking:
1) We aren't generally bothering doing defensive param checking in this lib (sure its a little lazy but its a common skip and rarely an issue) - is this something you feel like you want to see everywhere and if not should we just do it here?
2) Why the throw error instead of a cb(error) Guessing its because this is an input issue not a remote issue and you want to handle differently? 
3) Why check for name and options but not cb if this is full on param checking?

All for the use of the OSUtils.getArgsWithCallback when there are truly optional params but it seems like all values (including options) are actually required for all methods except list (where OSUtils.getArgsWithCallback seems appropriate...and maybe something we should implement across the board for filtering.)

thoughts??